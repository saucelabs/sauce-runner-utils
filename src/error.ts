export function errCode(obj: unknown): string | number {
  if (typeof obj !== 'object' || obj === null || !('code' in obj)) {
    return '';
  }
  if (typeof obj.code === 'number' || typeof obj.code === 'string') {
    return obj.code;
  }
  return '';
}
