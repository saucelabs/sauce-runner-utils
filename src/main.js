module.exports = require('./utils');
module.exports.saucectl = require('./saucectl');