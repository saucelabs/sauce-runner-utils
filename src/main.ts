/* istanbul ignore file */
// import utils from './utils';
// import saucectl from './saucectl';

// export default import from './utils';
// export saucectl { default as saucectl } from './saucectl';
