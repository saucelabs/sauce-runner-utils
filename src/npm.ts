import { spawn } from 'child_process';
import { lstat, rename, rm, writeFile } from 'fs/promises';
import { IHasNodePath } from './types';

const temporarilyMovedFiles: {[key: string]: string} = {
  'package.json': `.package.json-${process.pid}`,
  'package-lock.json': `.package-lock.json-${process.pid}`,
};

export default class NPM {
  static async removePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(name);
        await rename(name, replacement);
      } catch (e) {
        console.debug(`no ${name} to archive`);
      }
    }
  }

  static async restorePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(replacement);
        await rename(replacement, name);
      } catch (e) {
        console.debug(`no ${name} was archived`);
      }
    }
  }

  public static async install (nodePath: IHasNodePath, pkg: {[key: string]: string}) {
    await this.removePackageJson();
    await writeFile('package.json', JSON.stringify({
      dependencies: pkg,
    }));

    const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'install']);
    p.stdout.pipe(process.stdout);
    p.stderr.pipe(process.stderr);

    const exitPromise = new Promise((resolve) => {
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });

    const exitCode = await exitPromise;

    await rm('package.json', { force: true });
    await this.restorePackageJson();

    return exitCode;
  }

  public static configure (nodePath: IHasNodePath, cfg: {[key: string]: object | string | number | boolean | null }): Promise<number | null> {
    return new Promise((resolve) => {
      const args = Object.keys(cfg).map((k,) => `${k}=${cfg[k]}`);
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'config', 'set', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', () => {
        console.log('Finished');
        resolve(0);
      });
    });
  }

  public static rebuild (nodePath: IHasNodePath, ...args: string[]): Promise<number | null> {
    return new Promise((resolve) => {
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'rebuild', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });
  }
}

