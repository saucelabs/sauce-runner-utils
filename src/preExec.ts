import ChildProcess from 'child_process';
import os from 'os';

function spawnAsync (cmd: string, args: string[]) {
  return new Promise(function (resolve) {
    const proc = ChildProcess.spawn(cmd, args);
    proc.stdout.on('data', (data) => {
      process.stdout.write(data.toString());
    });
    proc.stderr.on('data', (data) => {
      process.stderr.write(data.toString());
    });
    proc.on('exit', function (exitCode) {
      resolve(exitCode);
    });
    proc.on('error', function (err) {
      console.log(`Unable to start command: ${err}`);
      resolve(1);
    });
  });
}

async function preExecRunner (preExecs: string[]) {
  const cmdInvoker = (os.platform() === 'win32') ? 'cmd' : 'sh';
  const cmdArg = (os.platform() === 'win32') ? '/C' : '-c';

  for (const command of preExecs) {
    console.log(`Executing pre-exec command: ${command}`);
    const exitCode = await spawnAsync(cmdInvoker, [cmdArg, command]);
    console.log('\n');

    if (exitCode !== 0) {
      return false;
    }
  }
  return true;
}

export async function run (suite: { preExec: string[] | undefined | null }, timeoutSec: number): Promise<boolean> {
  if (!suite.preExec) {
    return true;
  }

  let timeout;
  const timeoutPromise = new Promise((resolve) => {
    timeout = setTimeout(() => {
      console.error(`Pre-Exec timed out after ${timeoutSec} seconds`);
      resolve(false);
    }, timeoutSec * 1000);
  });
  const hasPassed: boolean = await Promise.race([timeoutPromise, preExecRunner(suite.preExec)]) as boolean;
  clearTimeout(timeout);
  return hasPassed;
}

export default run;
