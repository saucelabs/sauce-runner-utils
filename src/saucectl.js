const fs = require('fs');

// Store file containing job-details url.
// Path has to match the value of the Dockerfile label com.saucelabs.job-info !
const OUTPUT_FILE_PATH = '/tmp/output.json';

function exportValue (payload) {
  fs.writeFileSync(OUTPUT_FILE_PATH, JSON.stringify(payload));
}

function updateExportedValue (data) {
  let fileData;
  try {
    const st = fs.statSync(OUTPUT_FILE_PATH);
    if (st.isFile()) {
      fileData = JSON.parse(fs.readFileSync(OUTPUT_FILE_PATH)) || {};
    }
  } catch (e) {}
  fileData = { ...fileData, ...data };
  exportValue(fileData);
}

module.exports = {
  OUTPUT_FILE_PATH, exportValue, updateExportedValue,
};