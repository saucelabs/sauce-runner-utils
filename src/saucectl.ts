import { readFileSync, statSync, writeFileSync } from 'fs';

export function exportValue (filepath: string, payload: object | null) {
  writeFileSync(filepath, JSON.stringify(payload));
}

export function updateExportedValue (filepath: string, data: object | null) {
  let fileData;
  try {
    const st = statSync(filepath);
    if (st.isFile()) {
      fileData = JSON.parse(readFileSync(filepath).toString()) || {}; // FIXME: handle warning
    }
  } catch (e) {}
  fileData = { ...fileData, ...data };
  exportValue(filepath, fileData);
}
