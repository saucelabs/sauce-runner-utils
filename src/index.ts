import * as saucectl from './saucectl';
import * as npm from './npm';
import * as utils from './utils';
import {
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
} from './utils';

export {
  npm,
  saucectl,
  utils,

  // Exporting all to keep compatibility with previous API
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
};
