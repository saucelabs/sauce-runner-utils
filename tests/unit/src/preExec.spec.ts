import childProcess from 'child_process';
import events from 'events';
import stream from 'stream';
import os from 'os';
import { run } from '../../../src/preExec';

describe('preExec', function () {
  it('should return true when passing', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    setTimeout(() => {
      fakeProc.stdout?.emit('data', 'stdout-data');
      fakeProc.stderr?.emit('data', 'stderr-data');
      fakeProc.emit('exit', 0);
    }, 100);

    const hasPassed = await run({ preExec: ['npm install'] }, 10);

    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(true);
  });

  it('should return false when failing', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    setTimeout(() => {
      fakeProc.emit('exit', 1);
    }, 100);

    const hasPassed = await run({ preExec: ['npm install'] }, 10);

    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(false);
  });

  it('should return false when timeouting', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    const hasPassed = await run({ preExec: ['npm install'] }, 1);

    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(false);
  });

  it('should return false when failed to start', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    setTimeout(() => {
      fakeProc.emit('error', 'Unable to start command');
    }, 100);

    const hasPassed = await run({ preExec: ['npm install'] }, 10);

    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(false);
  });

  it('should return true when no preExecs', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');
    const hasPassed = await run({ preExec: undefined }, 1);
    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(true);
  });

  it('should return false when failed to start', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('linux');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    setTimeout(() => {
      fakeProc.emit('error', 'invalid start error');
    }, 100);

    const hasPassed = await run({ preExec: ['npm install'] }, 10);
    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(false);
  });

  it('should use cmd /C when running on windows', async function () {
    jest.spyOn(os, 'platform').mockReturnValue('win32');
    const spawnSpy = jest.spyOn(childProcess, 'spawn');

    const fakeProc = <childProcess.ChildProcess>new events.EventEmitter();
    fakeProc.stdin = new stream.Writable();
    fakeProc.stdout = <stream.Readable>new events.EventEmitter();
    fakeProc.stderr = <stream.Readable>new events.EventEmitter();
    spawnSpy.mockReturnValue(fakeProc);

    setTimeout(() => {
      fakeProc.emit('error', 'invalid start error');
    }, 100);

    const hasPassed = await run({ preExec: ['npm install'] }, 10);
    expect(spawnSpy.mock.calls).toMatchSnapshot();
    expect(hasPassed).toBe(false);
  });
});
