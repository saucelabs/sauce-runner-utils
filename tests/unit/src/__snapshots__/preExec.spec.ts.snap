// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preExec should return false when failed to start 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should return false when failed to start 2`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should return false when failing 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should return false when timeouting 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should return true when no preExecs 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should return true when passing 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
]
`;

exports[`preExec should use cmd /C when running on windows 1`] = `
Array [
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "sh",
    Array [
      "-c",
      "npm install",
    ],
  ],
  Array [
    "cmd",
    Array [
      "/C",
      "npm install",
    ],
  ],
]
`;
